GGIR 2.0

vincentvanhees opened the pull request:
**The part 5 specific changes:**
1. The changes are aimed to make better use of all the sleep information available from part 3 and 4.
2. New paragraph "Sleep and full day time-use analysis in GGIR" in the vignette which gives a full description of how the sleep and time-use detection currently work in GGIR. Note that I have moved output variable documentation for parts 2, 4 and 5 from tutorial to the vignette, such that the vignette becomes a more self-contained place for introductions on how to use GGIR and how to interpret the output. The GGIR function tutorial is then only needed for background information on function arguments.
3. The time series exported to csv in part 5 has been improved to only export the time series for the days that are considered valid in part 5, and now includes more detail than before, e.g. magnitude of acceleration and non-wear. In this way these files should be able to facilitate time series research,
4. As this will be a full new release I considered it a good moment to revise all the difficult to understand variables names in part 4 and 5 output. I did not change any function names or argument names to avoid too severe backward compatibility problems.
5. Part 5 now also stores full and cleaned output just like part 4. In that way the user has more opportunities to investigate why some days were excluded/included.
6. With all the additions the part 5 function code grew to over a 1000 lines. So, I split up the code in 7 new functions to ease navigation and debugging. I have not had time yet to write good unit tests for all of them, so that is still on my to do list.
7. Option added to force GGIR to ignore specific days in part 5 or to rely entirely on sleeplog for a specific night(s) or specific recording(s), see argument data_cleaning_file.

Edit: And it is now possible to aggregate the data from 5 seconds to 60 seconds in part 5 to facilitate some research I am involved in, this is turned off by default, so a normal user will not notice it.

**Other changes that will be part of version 2.0:**
- HDCZA algorithm expanded to be able to detect daysleepers
- Numunpack function moved back to c++ because .cwa file reading for AX3 became too slow.
- Various updates to visualreport (plot5 function) by Matt, I merged them in this pull request to make sure everything combines correctly.
- Part 2 report generation now able to handle changing variable count between milestone files data. Previously it behaved strange if the first recording in a dataset has insufficient valid days.
- Qwindow now able to handle smaller intervals (e.g. qwindow = seq(0,24,by=1)).
- GGIR now able to facilitate external function embedding, and this is explained in a separate package vignette.
- We now consistently refer to ID (not id) and calendar_date, spelling was 
  inconsistent before with sometimes id and then ID or calendardate and calender_date.
- Various improvements to qc plots generated by part 3, e.g. time line did not account for desiredtz but used local timezone and I have now added magnitude of acceleration to the plot.
- Expanded clipping detection: If any value in block is more than 150 percent the dynamic range then ignore entire block. This because there have been recordings where the signal had values of over a thousand g for for a few minutes. Such values were not picked up by the clipping detection because they did not span 80% of the window and as a result inflated the magnitude of acceleration for that recording.

vincentvanhees commented:
Hi @jhmigueles and @m-patterson 
In this pull request I would like to merge a lot of things I have been working on this year into the Master branch, from which GGIR version 2.0-0 can be created.

Would you mind doing the following:
- Check whether this version of GGIR affects your typical GGIR usage.
- Check whether you understand the new vignette [paragraph](https://github.com/wadpac/GGIR/blob/GGIR2.0/vignettes/GGIR.Rmd)  on "Sleep and full day time-use analysis in GGIR" (line 688), which explains how the sleep detection currently works. This hopefully provides you with an up to date overview of what GGIR does without having to go through all the code.
- Read the list of changes in this PR (above).
- If you want to dive deeper into all the changes I made then great, but I understand if you do not have time for that.

jhmigueles commented:
Testing now the branch in some files, everything looks ok, but I will further test it during the weekend.

I made some minor changes in the text (mainly fixed some typos).  Also, some comments in the file attached for you to consider. Feel free to consider or not the changes and comments.
[Sleep ggir.pdf](https://github.com/wadpac/GGIR/files/4530197/Sleep.ggir.pdf)

vincentvanhees commented:
I am now reversing the summary Matt, because the order was incorrect. It should be:

1. Get sustained inactivity Bouts (SIBs) on all data.
2. Define Sleep Period Time (SPT) window as the start of the first SIB and the end of the last SIB that overlap with a guiding method as detailed below.
3. Consider SIB during SPT window as sleep time.
4. Consider non-SIB during SPT window as nocturnal wake time.

Important point is that the guiding method is never considered the Sleep period time window. Only after looking at overlap with SIBs the Sleep Period Time window will be defined.

I will leave out the summary, because the summary is now quite short and I will work on the text to better convey this message.

jhmigueles commented on R/g.part5.savetimeseries.R:
```diff
@@ -0,0 +1,58 @@
+      # only doing this for RData output, because it would affect file size too much in csv,
+      # remember that this function can create many files: sample sizes times all combinations of thresholds.
+      mdat$timestamp = as.POSIXlt(mdat$timenum, origin="1970-01-01",tz="Europe/London") 
+      save(mdat,rawlevels_fname)
```
```suggestion
      save(mdat, file = rawlevels_fname)
```

vincentvanhees commented:
As you can see I have been trying to address issues with the build as a result of the R upgrade to 4.0.
This seems more complex to resolve, so I will start a separate branch to look at that rather than trying to fix it within this branch.

vincentvanhees commented:
- GGIR is now R 4.0 ready.
- I have now updated the part 5 behavioral class and subsequent variable names to be more intuitive.
- I have added to extra input argument to part 5 report:
  - includedaycrit.part5 to control inclusion criteria for the number of valid hours during the waking hours of a day.
   -  minimum_MM_length.part5 to control the minimum length of a day needed in part 5 when days are defined as MM.

In the upcoming days I will work a bit more on the documentation, e.g. clarify new variable names, and then I think the this branch is ready to be merged and turned into release 2.0.

vincentvanhees commented:
I forgot to mention in my previous message that I have now also decided to remove the behavioral category SIB during waking hours (daytime) from the part 5 because we hardly know what it means and the distinction with inactivity bouts is not clear. The underlying acceleration values will now be treated like any other segment of the data and most likely end up being classified as inactivity behavior.

Part 4 already provides an estimate of SIB during waking hours, which can facilitate methodological research around SIB.

vincentvanhees commented:
that was hopefully the final comment, I will test things one more time tomorrow and if there are no issues I will merge this and make the release (finally!)
